### PR TITLE
feat: messaging command - await response feature for send/reply

### LINKED ISSUE #2332: Messaging command - Await response feature for message send/reply

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- closes #2332

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

**Review by fqutishat — APPROVED** [review]

File: pkg/client/messaging/client.go [hunk 1]
```diff
@@ -141,108 +169,189 @@ func (c *Client) Services() []string {
...
 
-	return c.sendToDestination(msg, sendOpts)
+	return c.sendAndWaitForReply(action, didCommMsg.ID(), sendOpts.responseMsgType, sendOpts.timeout)
 }
 
 // Reply sends reply to existing message.
-func (c *Client) Reply(msg json.RawMessage, msgID string, startNewThread bool) error {
-	didCommMsg, err := service.ParseDIDCommMsgMap(msg)
+func (c *Client) Reply(msg json.RawMessage, msgID string, startNewThread bool,
+	waitForResponse string, timeout time.Duration) (json.RawMessage, error) {
```
Comment (troyronda) [inline pkg/client/messaging/client.go:206]: you should use a context not a time.Duration.

**Review by troyronda — DISMISSED** [review]

File: pkg/client/messaging/client.go [hunk 2]
```diff
@@ -95,6 +105,24 @@ func SendByDestination(destination *service.Destination) SendMessageOpions {
...
+// WaitForResponse option to set message response type.
+// Message reply will wait for the response of this message type and matching thread ID.
+func WaitForResponse(responseType string) SendMessageOpions {
+	return func(opts *sendMsgOpts) {
+		opts.responseMsgType = responseType
+	}
+}
+
+// ResponseTimeout is timeout waiting for response.
+func ResponseTimeout(timeout time.Duration) SendMessageOpions {
```
Comment (troyronda) [inline pkg/client/messaging/client.go:117]: should use go context style, not durations.

File: pkg/client/messaging/client.go [hunk 3]
```diff
@@ -141,108 +169,189 @@ func (c *Client) Services() []string {
...
-	if rqst.theirDID != "" {
-		var err error
+	return func() error {
+		return c.ctx.Messenger().SendToDestination(msg, base58.Encode(sigPubKey), dest)
+	}, nil
+}
 
-		dest, err = service.GetDestination(rqst.theirDID, c.ctx.VDRegistry())
+func (c *Client) sendAndWaitForReply(action messageDispatcher, thID string,
+	replyType string, timeout time.Duration) (json.RawMessage, error) {
```
Comment (troyronda) [inline pkg/client/messaging/client.go:281]: go context

File: pkg/client/messaging/client.go (hunk 1 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go:206]: ok, but JS & REST users will still be using timeout.

Comment (troyronda) [reply pkg/client/messaging/client.go:206]: That's okay in those layers. In a Go API, it needs to be a context.

Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

File: pkg/client/messaging/client.go (hunk 3 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

File: pkg/client/messaging/client.go (hunk 2 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

File: pkg/client/messaging/client.go [hunk 4]
```diff
@@ -95,6 +106,24 @@ func SendByDestination(destination *service.Destination) SendMessageOpions {
...
 
+// WaitForResponse option to set message response type.
+// Message reply will wait for the response of this message type and matching thread ID.
+func WaitForResponse(responseType string) SendMessageOpions {
+	return func(opts *sendMsgOpts) {
+		opts.responseMsgType = responseType
+	}
+}
+
+// WithContext is context for await for reply operation.
```
Comment (troyronda) [inline pkg/client/messaging/client.go:117]: A fresh context should be used for each operation. I'm a bit concerned about the option pattern. Can you clarify?

Comment (sudeshrshetty) [reply pkg/client/messaging/client.go:117]: we need context/timeout only when someone uses await reply option, if someone doesn't pass it then we use context with default timeout(20 seconds)

File: pkg/client/messaging/client.go [hunk 5]
```diff
@@ -141,108 +170,192 @@ func (c *Client) Services() []string {
...
+
+	if notificationCh != nil {
+		return waitForResponse(ctx, notificationCh)
+	}
+
+	return json.RawMessage{}, nil
+}
+
+func waitForResponse(ctx context.Context, notificationCh chan NotificationPayload) (json.RawMessage, error) {
+	if ctx == nil {
```
Comment (troyronda) [inline pkg/client/messaging/client.go:334]: It's a bit weird to see this check (and default setting) so far down in the call stack.

Comment (sudeshrshetty) [reply pkg/client/messaging/client.go:334]: I can move it up

Comment (troyronda) [reply pkg/client/messaging/client.go:334]: Yes move it up and couple it with the await reply option. 

Also no "defaulting" at this layer - make them pass you a context in the API layer.

Comment (sudeshrshetty) [reply pkg/client/messaging/client.go:334]: Forced caller to pass context whenever await reply option is used.

File: pkg/client/messaging/client.go [hunk 6]
```diff
@@ -95,6 +107,22 @@ func SendByDestination(destination *service.Destination) SendMessageOpions {
 	}
 }
 
+// WaitForResponse option to set message response type.
+// Message reply will wait for the response of this message type and matching thread ID.
+func WaitForResponse(ctx context.Context, responseType string) SendMessageOpions {
+	return func(opts *sendMsgOpts) {
+		if ctx == nil {
```
Comment (troyronda) [inline pkg/client/messaging/client.go:114]: Please remove.

File: pkg/client/messaging/client.go [hunk 7]
```diff
@@ -69,6 +73,14 @@ type sendMsgOpts struct {
...
 	// This param can be used to send messages outside connection.
 	destination *service.Destination
+
+	// Message type of the response for the message sent.
+	// If provided then messenger will wait for the response of this type after sending message.
+	responseMsgType string
+
+	// context for await reply operation.
+	ctx    context.Context
+	cancel context.CancelFunc
```
Comment (troyronda) [inline pkg/client/messaging/client.go:83]: Remove cancel.

File: pkg/client/messaging/client.go [hunk 8]
```diff
@@ -69,6 +73,14 @@ type sendMsgOpts struct {
 	// Destination is service endpoint destination.
 	// This param can be used to send messages outside connection.
 	destination *service.Destination
+
+	// Message type of the response for the message sent.
+	// If provided then messenger will wait for the response of this type after sending message.
+	responseMsgType string
+
+	// context for await reply operation.
+	ctx    context.Context
```
Comment (troyronda) [inline pkg/client/messaging/client.go:82]: This isn't a generic ctx - it is tied to your waitForResponse. Should name appropriately.

File: pkg/client/messaging/client.go (hunk 6 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

File: pkg/client/messaging/client.go (hunk 8 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

File: pkg/client/messaging/client.go (hunk 7 above)
Comment (sudeshrshetty) [reply pkg/client/messaging/client.go]: done

**Review by rolsonquadras — APPROVED** [review]

**Review by troyronda — APPROVED** [review]

**Pull request merged**